include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,427 @@
+                             const TensorShape& index_shape, const int32_t* index_data);
+
+  /// <summary>
+  /// Mutable data access
+  /// </summary>
+  class BlockSparseMutator {
+   public:
+    BlockSparseMutator(Tensor& values, Tensor& index) noexcept
+        : values_(values), index_(index) {}
+    Tensor& Values() noexcept { return values_; }
+    Tensor& Index() noexcept { return index_; }
+
+   private:
+    std::reference_wrapper<Tensor> values_;
+    std::reference_wrapper<Tensor> index_;
+  };
+
+  /// <summary>
+  /// Allocates memory for values and index and returns mutator so
+  /// data can be populated
+  /// </summary>
+  /// <param name="values_shape"></param>
+  /// <param name="index_shape"></param>
+  /// <returns></returns>
+  BlockSparseMutator MakeBlockSparseData(const TensorShape& values_shape, const TensorShape& index_shape);
+
+  /// <summary>
+  /// X-device copy. Destination tensor must have allocator set.
+  /// </summary>
+  /// <param name="data_transfer_manager"></param>
+  /// <param name="exec_q_id"></param>
+  /// <param name="dst_tensor"></param>
+  /// <returns></returns>
+  Status Copy(const DataTransferManager& data_transfer_manager, int exec_q_id, SparseTensor& dst_tensor) const;
+
+  /// <summary>
+  /// X-device copy. Destination tensor must have allocator set.
+  /// </summary>
+  /// <param name="dst_tensor"></param>
+  /// <returns></returns>
+  Status Copy(const IDataTransfer& data_transfer, SparseTensor& dst_tensor, int exec_q_id) const;
 
  private:
+  Status AllocateBuffer(int64_t buffer_size, size_t num_values);
+  void ReleaseBuffer();
+  void* IndicesStart(int64_t values_bytes);
+  const void* IndicesStart(int64_t values_bytes) const;
+  Status ValidateBlockSparseShapes(const TensorShape& values_shape, const TensorShape& index_shape) const;
+  void CopyStrings(const Tensor& src, Tensor& dst) const;
+
+  std::vector<int64_t> GetCooIndexDims(size_t values_count, size_t index_size) const;
+  void InitCooIndex(const TensorShape& index_shape, const int64_t* index_data);
+
+  void ValidateCsrIndices(size_t values_count, size_t inner_size, size_t outer_size) const;
+  void InitCsrIndices(size_t inner_size, const int64_t* inner, size_t outer_size, const int64_t* outer);
+
+  template <typename... T>
+  std::vector<std::reference_wrapper<Tensor>> MakeListNonConst(T&... t) {
+    return std::vector{std::ref(t)...};
+  }
+
+  template <typename... T>
+  std::vector<std::reference_wrapper<const Tensor>> MakeListConst(const T&... t) {
+    return std::vector{std::cref(t)...};
+  }
+
+  Status CopyData(const IDataTransfer& data_transfer,
+                  const std::vector<std::reference_wrapper<const Tensor>>& src,
+                  const std::vector<std::reference_wrapper<Tensor>>& dst);
+
+  SparseFormat format_;
+  TensorShape dense_shape_;
+  const PrimitiveDataTypeBase* ml_data_type_;
+  AllocatorPtr allocator_;
+  OrtMemoryInfo location_;
+  void* p_data_;
```

Comment:
Can you write a comment about what these data members mean? p_data_, values, etc?